File: apps/01-simple-chat/backend/golang/src/services/llm/gemini_service.go
package llm

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"time"
)

// GeminiService implements the LLMInterface for Google Gemini
type GeminiService struct {
	apiKey  string
	model   string
	baseURL string
	client  *http.Client
}

// NewGeminiService creates a new GeminiService
func NewGeminiService() (*GeminiService, error) {
	apiKey := os.Getenv("GEMINI_API_KEY")
	if apiKey == "" {
		return nil, fmt.Errorf("Gemini API key is not configured")
	}

	model := os.Getenv("GEMINI_MODEL")
	if model == "" {
		model = "gemini-2.0-flash"
	}

	client := &http.Client{
		Timeout: 30 * time.Second,
	}

	return &GeminiService{
		apiKey:  apiKey,
		model:   model,
		baseURL: "https://generativelanguage.googleapis.com/v1beta",
		client:  client,
	}, nil
}

// SendMessage sends a message to the Gemini API and returns the response
func (s *GeminiService) SendMessage(message string) (string, error) {
	requestBody := map[string]interface{}{
		"contents": []map[string]interface{}{
			{
				"parts": []map[string]string{
					{
						"text": message,
					},
				},
			},
		},
		"generationConfig": map[string]interface{}{
			"temperature":     0.7,
			"topK":           40,
			"topP":           0.95,
			"maxOutputTokens": 1024,
		},
	}

	jsonData, err := json.Marshal(requestBody)
	if err != nil {
		return "", fmt.Errorf("error marshaling request: %w", err)
	}

	url := fmt.Sprintf("%s/models/%s:generateContent?key=%s", s.baseURL, s.model, s.apiKey)
	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
	if err != nil {
		return "", fmt.Errorf("error creating request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")

	resp, err := s.client.Do(req)
	if err != nil {
		return "", fmt.Errorf("error making request to Gemini API: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("Gemini API error (status %d): %s", resp.StatusCode, string(body))
	}

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", fmt.Errorf("error decoding response: %w", err)
	}

	candidates, ok := result["candidates"].([]interface{})
	if !ok || len(candidates) == 0 {
		return "", fmt.Errorf("no candidates returned from Gemini API")
	}

	candidate, ok := candidates[0].(map[string]interface{})
	if !ok {
		return "", fmt.Errorf("invalid candidate format from Gemini API")
	}

	content, ok := candidate["content"].(map[string]interface{})
	if !ok {
		return "", fmt.Errorf("invalid content format from Gemini API")
	}

	parts, ok := content["parts"].([]interface{})
	if !ok || len(parts) == 0 {
		return "", fmt.Errorf("no parts returned from Gemini API")
	}

	part, ok := parts[0].(map[string]interface{})
	if !ok {
		return "", fmt.Errorf("invalid part format from Gemini API")
	}

	text, ok := part["text"].(string)
	if !ok {
		return "", fmt.Errorf("invalid text format from Gemini API")
	}

	return text, nil
}

// GetProvider returns the name of the LLM provider
func (s *GeminiService) GetProvider() string {
	return "Google Gemini"
}

// GetModel returns the model name used by the LLM provider
func (s *GeminiService) GetModel() string {
	return s.model
} 

File: apps/01-simple-chat/backend/python/src/services/llm/gemini_service.py
import os
import httpx
from .llm_interface import LLMInterface

class GeminiService(LLMInterface):
    """Gemini service implementation"""
    
    def __init__(self):
        self.api_key = os.getenv("GEMINI_API_KEY")
        self.model = os.getenv("GEMINI_MODEL", "gemini-2.0-flash")
        self.base_url = "https://generativelanguage.googleapis.com/v1beta"
        self.client = httpx.AsyncClient(
            base_url=self.base_url,
            timeout=30.0
        )
        
        if not self.api_key:
            raise Exception("Gemini API key is not configured")
    
    async def send_message(self, message: str) -> str:
        """Send a message to the Gemini API and get a response"""
        try:
            url = f"/models/{self.model}:generateContent?key={self.api_key}"
            
            response = await self.client.post(
                url,
                json={
                    "contents": [
                        {
                            "parts": [
                                {
                                    "text": message
                                }
                            ]
                        }
                    ],
                    "generationConfig": {
                        "temperature": 0.7,
                        "topK": 40,
                        "topP": 0.95,
                        "maxOutputTokens": 1024,
                    }
                }
            )
            
            if response.status_code != 200:
                raise Exception(f"Gemini API error: {response.text}")
            
            data = response.json()
            
            if not data.get("candidates") or len(data["candidates"]) == 0:
                raise Exception(f"No candidates returned from Gemini API: {data}")
            
            if not data["candidates"][0].get("content") or not data["candidates"][0]["content"].get("parts"):
                raise Exception(f"Invalid response format from Gemini API: {data}")
            
            return data["candidates"][0]["content"]["parts"][0]["text"]
        except Exception as e:
            raise Exception(f"Error communicating with Gemini API: {str(e)}")
    
    def get_provider(self) -> str:
        """Get the name of the LLM provider"""
        return "Google Gemini"
    
    def get_model(self) -> str:
        """Get the model name used by the LLM provider"""
        return self.model 